src/cpp/rtps/messages/MessageReceiver.cpp

```diff
@@ -662,27 +662,50 @@ bool MessageReceiver::proc_Submsg_Data(
         ch.sourceTimestamp = timestamp_;
     }
 
-#if HAVE_SECURITY
-    if (first_reader->getAttributes().security_attributes().is_payload_protected)
-    {
-        if (participant_->security_manager().decode_serialized_payload(ch.serializedPayload, crypto_payload_,
-                first_reader->getGuid(), ch.writerGUID))
-        {
-            ch.serializedPayload.data = crypto_payload_.data;
-            ch.serializedPayload.length = crypto_payload_.length;
-        }
-    }
-#endif  // HAVE_SECURITY
-
     logInfo(RTPS_MSG_IN, IDSTRING "from Writer " << ch.writerGUID << "; possible RTPSReader entities: " <<
             associated_readers_.size());
 
-    //Look for the correct reader to add the change
-    findAllReaders(readerID,
-            [&ch](RTPSReader* reader)
+#if HAVE_SECURITY
+    auto process_message = [&ch, this](RTPSReader* reader)
             {
-                reader->processDataMsg(&ch);
-            });
+                if (reader->matched_writer_is_matched(ch.writerGUID))
+                {
+                    bool is_decoded = false;
+                    if (reader->getAttributes().security_attributes().is_payload_protected)
+                    {
+                        if (participant_->security_manager().decode_serialized_payload(ch.serializedPayload,
+                                crypto_payload_, reader->getGuid(), ch.writerGUID))
+                        {
+                            std::swap(ch.serializedPayload.data, crypto_payload_.data);
+                            std::swap(ch.serializedPayload.length, crypto_payload_.length);
+                            is_decoded = true;
+                        }
+                        else
+                        {
+                            return;
+                        }
+                    }
+                    reader->processDataMsg(&ch);
+
+                    if (is_decoded)
+                    {
+                        std::swap(ch.serializedPayload.data, crypto_payload_.data);
+                        std::swap(ch.serializedPayload.length, crypto_payload_.length);
+                    }
+                }
+            };
+#else
+    auto process_message = [&ch](RTPSReader* reader)
+            {
+                if (reader->matched_writer_is_matched(ch.writerGUID))
+                {
+                    reader->processDataMsg(&ch);
+                }
+            };
+#endif  // HAVE_SECURITY
```

Comment:
Some readers allow the reception of data from non-matched writers. If that is the case, they will never have `is_payload_protected` set.

So I think the correct solution is as follows:
* When built without security, always call `processDataMsg` directly
* When built with security:
    * If `is_payload_protected` is not set, call `processDataMsg`
    * Otherwise, if writer is not matched, abort lambda
    * Finally, try to decode the payload